\documentclass{comjnl}
\usepackage{ragged2e}
\usepackage{amsmath}
\usepackage{amssymb}

%----
  \setlength\textheight{220mm}
%----

\input paper-seb-macros.tex
\input generated/info-for-main.tex
\input generated/page-lengths.tex

\long\def\ignore#1{}

\title{\mytitle} 
\author{Harold Thimbleby}
\shortauthors{Harold Thimbleby}
\affiliation{See Change Fellow in Digital Health, Swansea University}
\email{harold@thimbleby.net}

%\subject{Computational science (bioinformatics,~medical~computing,~etc)}

\keywords{Computational Science; Software Engineering; Reproducibility; Scientific scrutiny; Reproducible Analytic Pipeline (RAP \& \RAPstar\hskip -1pt)}

\received{30 April 2022}
%\revised{--}

%\begin{tabular}{@{}ll}Current Address: &62 Cyncoed Road, Cardiff, CF23 5SH, Wales\\
%&\texttt{harold@thimbleby.net}\\
%&ORCID 0000-0003-2222-4243
%\end{tabular}

%\orcid{0000-0003-2222-4243}

\begin{document}

\begin{abstract}
\justifying
%\color{red}
%This is a placeholder for the revisions that are almost complete. I have uploaded it now in case the computer system would lose all record of the submission --- I have emailed to ask for a short extension, but with the offices shut, there is no chance of any negotiation. Thanks for your understanding.
%\color{black}\vskip .2cm

\noindent
Scientists rarely assure the structure and quality of the code they rely on, and they rarely make it accessible for scrutiny. Even if available, scientists rarely provide adequate documentation to understand or use their code reliably. When code used in science avoids adequate scrutiny published science will be unreliable, because results --- data, graphs, images, inferences, etc --- generated by the code will themselves be unreliable.  

\hskip 1em This paper justifies and proposes ways to improve research and science relying on code:

%\emph{Problems:} Assumptions in scientific code are hard to scrutinize as they are rarely made explicit, even when code is made available. Both algorithms and code have bugs, unknown and accidental assumptions that have unwanted effects. Code is fallible, so any interpretation that relies on code is also fallible. When the code is not clearly structured and published with adequate documentation, the code cannot be usefully scrutinized. In turn, scientific claims cannot be properly scrutinized.

%\emph{Methodology:} From the perspective of Software Engineering, this paper critiques the quality (and accessibility) of coding in computational science, particularly as relied on in leading COVID-19 pandemic research driving international public health interventions. 
%The paper reports a pilot survey of peer-reviewed computational modeling papers ($N=\the\dataN$) published in leading scientific journals.

%Code can be improved using Software Engineering. This paper argues for specific solutions:

%\emph{Results:} %Journals do not always have a code policy, and many have a relaxed approach to code disclosure. 

%\noindent\emph{Solutions:} 

\raggedright
\newdimen \mywidth \mywidth=\textwidth
\advance \mywidth by -3em

\def\indented#1{\vskip 1ex
\parshape 2 0em \textwidth 2em \mywidth 
\noindent 
\hbox to 2em{ #1.\hfill}\ignorespaces}

\indented 1
Professional Software Engineers can help, particularly in critical fields such as public health, climate change, and energy. 

\indented 3
``Software Engineering Boards,'' analogous to Ethics or Institutional Review Boards, should be instigated and used.

\indented 4
The Reproducible Analytic Pipeline (RAP) methodology can be generalized to cover code and Software Engineering methodologies, in a generalization this paper introduces called \RAPstarp. \RAPstar\ (or comparable interventions) could be supported and encouraged in journal, conference, and funding body policies.

\vskip 1ex 
\justifying
\noindent
The key point is that as code is now an inextricable part of science it should be supported by competent Software Engineering input, analogously to statistics being properly supported by critical and competent statistics input. The \supplement\ provides a summary of Software Engineering best practice relevant to scientific research, including further suggestions for \RAPstar\ processes. 

%\begin{quote}
\def\seprule{\vskip 2ex\hrule\vskip 2ex}
\seprule
\raggedright
\def\zq{\setbox0=\hbox{``}\hskip -0\wd0\copy0}
\zq Science is what we understand well enough to explain to a computer.''\\
\hfill {\rm Donald E. Knuth in $A=B$ \cite{a=b}}\\

\zq I have to write to discover what I am doing.''\\
\hfill {\rm Flannery O'Connor, quoted in \emph{Write for your life\/} \cite{write-for-life}}

\zq Criticism is the mother of methodology.''\\
\hfill {\rm Robert P. Abelson in \emph{Statistics as Principled Argument\/} \cite{abelson}}

\zq \hangindent \wd0 From its earliest times, science has operated by being open and transparent about methods and evidence, regardless of which technology has been in vogue.''\\
\hfill {\rm Editorial in \emph{Nature\/} \cite{nature-chatgpt}}

\seprule
%\end{quote}
\end{abstract}
\maketitle

%\section*{Author summary}
%
%%\includegraphics[width=1in,height=1.25in,clip,keepaspectratio]{HT}}]
%Harold Thimbleby PhD, FRCP (Edinburgh), Hon.\ FRSA, Hon.\ FRCP is See Change Fellow in Digital Health at Swansea University, Wales. His research focuses on human error and computer system design, particularly for healthcare. 
%%In addition to over 340 peer-reviewed and 188 invited publications, Harold has written several books, including \emph{Press On\/} (MIT Press, 2007), which was winner of the American Association of Publishers best book in computer science award.
%Harold won the British Computer Society Wilkes Medal. He is emeritus Gresham Professor of Geometry (a chair founded in 1597), and has been a Royal Society-Leverhulme Trust Senior Research Fellow and a Royal Society-Wolfson Research Merit Award holder. His latest book is  \emph{Fix IT: How to see and solve the problems of digital healthcare\/} \cite{fixit}.  See his web site, \texttt{www.harold.thimbleby.net}, for more details.

%\noindent\emph{This paper evidences widespread problems arising from scientific papers relying on inadequately defined computer and inaccessible code, which undermines peer-review and replication, and undermines scientific quality generally. The risks vary, depending on how critical a paper's claims may be: ranging from superficial for opinion papers, to serious for papers in domains such as medicine, public health, or policy. This paper contrasts the situation with the mature approach to using statistics, and goes on to make proposals to help address the problem. The proposals are feasible, and presented in the hope of motivating their (or better solutions) being implemented, tested, and refined.}

\section{Introduction}
Unreliable, often unstated and unexplored, code and computational dependencies (including using AI or ML) in science are widespread. Furthermore, code is rarely published or made accessible in a useful form, nor is it professionally scrutinized. Too often, therefore, code and results claimed do not contribute to reliable science, and do not support replication or reproduction on which future science can be firmly based.   

\begin{table}
\vskip 1.2ex % align top of columns!
\tableofcontents
\vskip 3mm \hrule
\end{table}

This paper makes an explicit analogy between the use of statistics, which has clear standards for reliable use and presentation, and the use of code. Like statistics, code is often relied on to support key results in scientific publications, yet code is generally informal, inaccessible and incorrect. Just as sound experimental methods and sound statistics generally rely on professional specialist input, it is argued here that good use of code must rely on professional Software Engineering input \cite{sommerville,knight}, and more strategically on Computational Thinking \cite{computational-thinking-wing,computational-thinking} --- an accessible form of Software Engineering that consciously applies the ideas universally, far more widely than just to software and coding.

\begin{table*}
\begin{center}\sf\normalsize
\begin{tabular}{|rl|} \hline
\the\numberOfJournals&Journals\\
\the\dataN&Papers:\\
\tabularJournalBreakdown
\the\countAuthors&Published authors\\
\the\totalPages&Published journal pages\\
July 2020&Sample month\\ \hline
\end{tabular}
\end{center}
\caption{Overview of peer-reviewed paper sample, broken down further in Table \ref{table-summary}. Survey methodology and data is provided in the \supplement. (The survey does not include the motivating papers \cite{ICmodel,avianFluModel,originalICmodel}, none of which provide code or code summaries; see section \ref{section-pandemic-modeling}.)}
\label{table-overview}
\end{table*}

\begin{table*}
\begin{center}\sf\normalsize
\input generated/summary-table.tex
\end{center}
\caption{Breakdown of pilot survey of peer-reviewed science papers relying on code. }
\label{table-summary}
\end{table*}

\begin{table*}
\newdimen\quarterwidth
\newdimen\separation
\separation=3mm
\quarterwidth=\textwidth
\advance\quarterwidth by -4\separation
\divide\quarterwidth by 4
\noindent{\sf\begin{tabular}{@{}c|c|c|c@{}}
\bfseries Level 0&
\bfseries Level 1&
\bfseries Level 2&
\bfseries Level 3 \\ \hline
\multicolumn{1}{@{}p{\quarterwidth}}{Journal encourages code sharing -- or says nothing.}&
\multicolumn{1}{|p{\quarterwidth}}{Article states whether code is available and, if so, where to access them.}&
\multicolumn{1}{|p{\quarterwidth}}{Code must be posted to a trusted repository. Exceptions must be identified at article submission.}&
\multicolumn{1}{|p{\quarterwidth}@{}}{Code must be posted to a trusted repository, and reported analyses will be reproduced independently before publication.}
\end{tabular}}
\caption{The TOP committee's recommended levels for journal article code transparency. Level 0 is provided for a comparison that does not meet any TOP requirements. Concerns about the interpretation of ``reproduced independently,'' as required at level 4, are raised in section \ref{reproducibility-concerns}.}
\label{TOPtable}
\end{table*}

This paper was initially motivated by concerns about the poor quality of code used in high-profile epidemiology research, because of its significance for informing and driving public health responses to the Covid pandemic. A pilot survey implies that such problems are ubiquitous and by no means limited to epidemiology: see the tables \ref{table-overview} \& \ref{table-summary} for a summary of the sample, and see the \supplement\ for further details of the survey. In the survey, \emph{no\/} papers claimed or provided evidence that their code was adequately tested or rigorously developed; \emph{none\/} used methodologies like RAP or \RAPstar\ (described below). Only one paper mentioned any Software Engineering methods, albeit without technical details. 

In the survey sample 81\% of papers were published in journals that have code policies (which themselves are weak), but 42\% of surveyed papers published in those journals breached their own policies. One paper declared it had accessible code, but the relevant repository was and still remains empty. The findings are comparable to problems increasingly recognized for data and data access (reviewed in section \ref{related-work}): code problems form part of the \emph{reproducibility crisis\/} \citeeg{reproducibility-crisis,rescience}.

This paper therefore argues that code should be developed and discussed in a professional, rigorous, and supportive environment that facilitates quality science with clear presentation and appropriately rigorous scrutiny of code. Its main contribution is to suggest straightforward ways to enable this. The proposals may not be ``the'' right or best ways, but it is hoped the case studies and arguments presented here persuade readers that the proposals are at least a productive way to start pointing in the right direction, and inform raising the profile and constructively debating the issues more widely. 

An extensive online \supplement\ appendix to this paper provides additional resources, including brief details of many Software Engineering practices relevant to supporting quality science. The supplement will be of particular interest to research software engineers supporting non-software-specialist scientists.

\section{Background}
The discoveries and inventions of scientific technologies and instruments like microscopes, telescopes, and X-rays, drive and expand the sciences. There are fascinating periods as new ideas and science unify; for example, thermometers could not measure temperature in any meaningful way until the science was mature. For over a century, there was no agreement on definitions of temperature, how to calibrate thermometers, or what units they measured in. 

Paradoxically the science could not mature until there was consensus in scientific methodologies for thermometry, and having a consensus in turn depended on reproducible thermometer measurements and a thorough understanding of the science, including all the confounding factors that were being misunderstood \cite{temperature}.\footnote{One example: if the volume of mercury is chosen to measure temperature, a confounding factor is that the volume of the container measuring the mercury volume also increases with temperature (but at a different rate), so the volume measurement is inaccurate.} 
Science matured from no quantitative interest in temperature, through a complex process of hand-in-hand theoretical-and-technical maturation, until today, when we have robust off-the-shelf instruments that measure temperature in reliable, repeatable, internationally standardized units, following international quality standards.

Computers are a unique, new technology, far more flexible and challenging than thermometers. Understanding computers and integrating them into science is likely to be harder than the development of modern thermometry. Computation not only expands science's paradigms and supports new discoveries (particularly with AI), but it also \emph{does\/} new science --- almost all modern laboratory instrumentation, including thermometers, is heavily computerized. Pushing the boundaries of science, then, involves pushing the boundaries of computer science. The synergy runs deep: for instance, while particle physics relies on powerful supercomputers, quantum physics itself is developing more powerful quantum computing.

``Computational science'' has come to mean a particular style of science based on developing and using explicit computational models, but, really, \emph{all\/} of science is now computational in this sense. Computational science is not just restricted to specialized fields like computational chemistry, genomics, big data~\ldots\ In all fields of science, computation is used at every step, from calculations of course, through note taking, sound and image processing, literature searches, analysis and statistics, correspondence with co-authors and editors, through to typesetting,  distributing and archiving the final publications. All areas of science are being profoundly computerized. Furthermore, developments in computer science themselves drive science such that earlier science is even becoming obsolete as the computer technology moves on \cite{form}.

\subsection{Related code quality concerns}\label{related-work}
Publishing high quality computer code has been strongly advocated since the earliest times, such as the \emph{Communications of the ACM\/} in its first issue in its first volume published in 1958, where it outlined its algorithm publication policy. The new policy was illustrated with a square root algorithm \cite{acm-algorithms}. However, publishing code in the computer science literature is distinct from publishing high quality general science that depends on code, which is the concern of the present paper. Of course, as a special case, Computer Science too can also benefit from improving ways to reliably use code.

In almost all published science, the code it relies on is taken for granted, just as in routine chemistry the quality of the glassware is not at issue. While chemists are trained in reliable methodologies, so taking quality glassware for granted is reasonable. Code is a newer innovation. Reliable code is a current research programme in its own right, and has resulted in calls for a Grand Challenge research effort \cite{hoare-correctness}. Inevitably, because of these reasons taken collectively, much published science depends on unreliable code that is not explicitly discussed, was not peer-reviewed, and is not open to scrutiny, reproduction, or reuse by the scientific community. 

A study of 863,878 Python-coded Jupyter notebooks \cite{jupyter-study} found a 76\% failure rate for code to complete execution successfully. Trisovic \emph{et al.}\ \cite{trisovic} performed a  study of 9,000 research codes written in the language R on Dataverse, an open-source repository maintained by Harvard University's Institute for Quantitative Social Sciences. They found a comparable result that 74\% of the code files analyzed failed. These results are consistent with this paper's findings, as summarized in tables \ref{table-overview} \& \ref{table-summary}. 

The authors of \cite{jupyter-study} make technical recommendations to improve reproducibility, such as ``Abstract code into functions, classes, and modules and test them.'' More generally, the authors of \cite{trisovic} recommend establishing Working Groups to support reproducible research --- and in fact Dataverse has already done so. These ideas may be compared to the present paper's proposal of Software Engineering Boards (SEBs), as discussed in detail in section \ref{summary}, supported by more specific suggestions in the \supplement.

These concerns about code are part of the reproducibility crisis for science generally \cite{machine-learning-reproducibility,no-raw-data,data-access,reproducible-manifesto,reproducibility-crisis,reproducible-manifesto}. Concern has led to a new journal \emph{ReScience~C\/} to explore and encourage the explicit replication of previously published research \cite{rescience}; furthermore, \emph{ReScience~C\/} promotes new and open-source implementations in order to ensure that the original research is reproducible. However, most of its published concern focuses on data rather than code. 

To start to address code quality issues, the Transparency and Openness Promotion Committee first met in 2014, and has since been promoting Transparency and Openness Promotion, TOP, starting with journal publication policies \cite{TOP}. TOP covers citation standards, replication standards, and code standards amongst others. TOP recommends levels of compliance to their recommendations, where level 0 does not meet the standard, and levels 1 to 3 are increasingly stringent. The TOP levels for code are shown in table \ref{TOPtable}. The TOP standards  continue to develop, and are now maintained on a wiki at \url{osf.io/9f6gx/wiki/Guidelines} \cite{TOP-wiki}. Note that TOP aligns with but is a stricter approach than the Findable, Accessible, Interoperable and Reusable (FAIR) initiative, which is critiqued in section \ref{critique-fair}.

As the present paper argues, developing quality code is widely under-appreciated, which leads to a vicious cycle of lack of acknowledgement, invisibility, and being unable to recruit adequately competent coders (section \ref{section-vicious-cycles}). The term \emph{research software engineer\/} was coined in 2012 to help address this problem, and to stimulate thinking about researchers' career paths. The Society of Research Software Engineering (\url{https://society-rse.org}) has been established to further promote research software engineer interests.  

There is no shortage of computational tools available to help address the problems. However, it should be noted that such tools are not a panacea, as the study \cite{jupyter-study} cited above makes clear. This suggests that human support for improving coding and reproduction quality, perhaps in the form of Working Groups or Boards, as this paper suggests, will be important.

In addition to unintentional problems with code quality and reproducibility, scientific misconduct occurs when the outcome is intentional. While pure plagiarism, which is misconduct, generally does not affect the quality of reported science, when data or code is fraudulently manipulated to have realistic properties, the results are likely to be destructive. The Wakefield MMR fraud linking vaccines and autism published in \emph{The Lancet\/} took 12 years before it was retracted, and has been extraordinarily destructive \cite{MMR-fraud}. 

A recent meta-analysis of surveys of scientific misconduct estimates that nearly 2\% of scientists at least once fabricated, falsified, or modified results, and over a third undertook other questionable research practices \cite{misconduct-issues}. The meta-analysis qualifies the figures carefully, but these are alarming rates with any qualifications; indeed, the authors suggest that as misconduct is a sensitive issue, the rates are likely to be under-estimates.\footnote{The author's survey of scientists publishing in the \emph{Journal of Machine Learning\/} had comparable results \cite{machine-learning-reproducibility}.} 

While technical solutions like using AI may help, it is notable that many misconduct issues can be detected and constructively managed prior to publication using the same methods as will improve research reproducibility, as discussed throughout the present paper.

\subsection{Computable papers}\label{computable-papers}
``Electronic lab notebooks'' (ELNs) \cite{elns}, emphasize computer tools that specifically support laboratory notebook authoring and editing. In contrast, ``computable papers'' aim to support the scientific paper authoring process: the emphasis is that papers should produce faithful results from code embedded in (or easily accessible from) the text of the paper. 

If, for example, HTML is being used to prepare a computational paper, it could include Javascript code like \texttt{<script> document.write(responses.total) </script>}, which would insert the result of running that Javascript code into the paper, such as the number into a sentence in an empirical paper, like ``We collected data and obtained 754 responses.'' 

Systems like \LaTeX\ (which was used for the current paper) can combine advanced typesetting with basic calculations (e.g., here we calculate $10!=
\def\factorial#1{\ffactorial{#1}1}
\newcount\t
\def\ffactorial#1#2{%
\ifnum#1=0 #2\else
\ffactorial{\the\numexpr#1-1\relax}{\the\numexpr#1*#2\relax}%
\fi}
\factorial{10}$ by writing \LaTeX\ code \emph{in\/} this paper). More practically, a separate program can generate a file of \LaTeX\ definitions for data, tables, and cross-references, etc, as needed for a paper. 

\newdimen\boxwidth
\boxwidth = \columnwidth
\advance\boxwidth by - 1em
\noindent\framebox[\columnwidth][r]{\parbox{\boxwidth}{\noindent\sf For the present paper, as a concrete example of this approach, the pilot survey analyzed \textbf{\the\dataN} papers with \textbf{\the\countAuthors} authors. The paper \textbf{\cite{ICmodel}}, discussed below in section \textbf{\ref{section-pandemic-modeling}}, relied on code composed of \textbf{\covidsimfiles} files, with over \textbf{\covidsimkLOC\ thousand} lines of code (how this compares with files in the survey is summarized in table \textbf{\ref{supplementaryTableRepoSummary}} in the \supplement).}}

In the ``old days'' such numbers (\the\dataN, \the\countAuthors, etc) quoted above would have been manually worked out, then eyeballed and typed up by the authors. Instead, in this paper, those numbers (and many more) were computed automatically. Furthermore, they update automatically when the data or calculations change. They were inserted into the text of this paper automatically, and are auditable back to their sources.  

The idea is easy to implement. Code can generate text like \texttt{\textbackslash newcommand\{\textbackslash numberOfAuthors\}\{\the\countAuthors\}} (where the \the\countAuthors\ is the calculated number) save that line of text to a file, which is then input into the paper where it is typeset in the normal way. Then, when and wherever the authors write the name \texttt{\textbackslash numberOfAuthors} in their paper's text, the typeset paper says \the\countAuthors, or whatever the correct value is at the time. Some easy \LaTeX\ coding can then present the numbers in the publisher's preferred style, such as \numbertoname 0, \numbertoname 1, \ldots\ \numbertoname 9, \numbertoname{10}, \numbertoname{11} \ldots\ 1,000 etc, as done (for example) in this paper's footnote~\numbernamesexamplefootnote. 

Rather than use quirky languages like \LaTeX\ computable papers more often build on general-purpose programming languages. Python and R, for instance, have powerful features for processing data, plotting graphs, doing statistics, and more.

The key concept in computable papers is that as the authors of a paper collect or revise data or calculations or otherwise update it, the results in the paper update automatically, and all the statistics, graphs, and analysis reported in the paper remain correct --- \emph{with no further work from the authors}. Indeed, if an author corrects a mistake, the correction will apply automatically to all future edits.

However, there is no structure to using general-purpose systems like HTML, \LaTeX, or Rmarkdown. Many authors therefore prefer a more structured approach imposed by tools designed for the purpose. A taut review is Perkel \cite{perkel-review}, but here four tools,  WEB, \emph{Mathematica}, Jupyter, and knitr, will serve to sample the variety of approaches that are available:

\begin{itemize}\raggedright

\item
WEB is the earliest tool reviewed here. WEB was developed by Donald Knuth in 1984 \cite{LP,LP-book} as a batch (non-interactive) tool to support his then radical new concept of \emph{literate programming}. The idea was to facilitate programmers write literate documentation for their code. WEB combines a sequential documentation file with code that can be presented in any order, thus overcoming the problem that the best explanation of a program is not necessarily written in the same order as the code it explains. The original WEB allowed Pascal programs to be documented in \TeX, but many variants of WEB have since been developed that are more flexible in the systems they support. 

\hskip 1em WEB documents an entire program, but there are variants such as relit \cite{relit} that allow arbitrary parts of programs to be documented, and hence are useful for normal scientific papers that need to explain algorithms, but do not need to show or explain the entire code.

\hskip 1em In contrast to the other tools reviewed here, literate programming is intended to produce high quality publications \emph{about\/} code, rather than publications \emph{using\/} the code, inserting output, such as graphs, generated by running it.

\item
\emph{Mathematica\/} is one of the earliest fully interactive notebook tools for computational papers. Notebooks were developed by Theodore Gray in 1988 \cite{notebook-patent}. Notebooks consist of a collection of ``cells.'' Some notebook cells are labelled as text or as section headings, but if a cell is labelled as code, it can be run and it will normally generate a new cell following it as its output. The new cell can be numbers, tables, mathematics, a plot, an image, or even more text --- the arbitrary output of running code, in fact. Moreover, a notebook itself is a \emph{Mathematica\/} expression, so it --- the paper itself --- can be analyzed or manipulated by code in any way. In particular, the entire notebook structure and contents can be checked for consistency and correctness in any way the author chooses.

\hskip 1em A \emph{Mathematica\/} notebook can be published directly as a paper, but some code might be distracting. Typically the author therefore hides some or all code cells that are irrelevant to the narrative of the paper, but which nonetheless were required to generate the results presented. 

\hskip 1em The user manual for \emph{Mathematica\/} itself was written as a \emph{Mathematica\/} notebook, which ensured all its examples actually worked --- and probably helped ensure the correctness of the \emph{Mathematica\/} code used behind the scenes (see section \ref{paper-as-lab}). The book is now the largest example of software documentation in existence: in its latest edition it runs to over 10,000 pages. 

\item
Jupyter was developed by Fernando P{\'e}rez and Brian Granger \cite{jupyter-history}, and takes a similar approach to \emph{Mathematica\/}, but is open-source and not closely integrated with any particular programming language, as \emph{Mathematica\/} is. Jupyter can be installed on a PC or run over the web. 

\hskip 1em Jupyter is both an authoring tool and a framework on which to build other tools: thus Google's colaboratory is built on top of Jupyter, using it as a foundation but making stylistic changes, including providing free computational resources. 

\hskip 1em Many extensive examples of using Jupyter notebooks (and other good practice, such as using repositories) to support large scale science projects can be found at the Gravitational Wave Open Science Center at \url{www.gw-openscience.org}. 

\item
knitr \cite{knitr} is a powerful culmination of a variety of tools, Pweave, Sweave, and ideas from literate programming. Knitr combines a markdown document with R code, and is a more powerful approach than the analogous, and perhaps more familiar, HTML+Javascript example shown above to motivate this section.
\end{itemize}

\noindent
Thimbleby \cite{spec-led-design} is a 1999 example of a peer-reviewed paper (about user interface design) written as a \emph{Mathematica\/} notebook, which makes the point that a distinctive feature of its methodology is that the \emph{Mathematica\/} notebook creates a fully inspectable and replicable process. The notebook is available on the author's web site; it can be checked by others, or easily extended or repurposed to support new research --- and it still works 
{\newcount \howlong \howlong = \year
\advance \howlong by -1999
\the\howlong}
years later.

There are tools to make using code more convenient and more reproducible, as this section briefly reviewed, but they are rarely used or used haphazardly, as the next section shows.

\subsection{RAP: Reproducible Analytical Pipelines}\label{RAP-section}
Writing a paper typically starts in a word processor (such as Microsoft Word), sketching an outline, writing boiler-plate text (such as the authors' names and standard section headings), and then gradually building up the evidence base (including citing the literature) that the paper relies on. The process will be concurrent with many other activities --- grant writing, writing up lab books, negotiating authorship, protecting IP, workshops, finding publication outlets, and so on.

Table \ref{fig-pipeline} illustrates the core pipeline of how experiments and data are used to provide information on which analysis and calculations are based, the results of which are then edited into a paper.

\begin{table*}
{\sf
\begin{center}
\def\drop#1{\setbox0=\hbox{\lower .5em\hbox{#1}}%
\ht0=0em
\dp0=0em
\copy0}
\def\lowarrow{\drop{$\rightarrow$}}

\def\bigBracket{$
	\setbox0=\hbox{\lower 1.2ex\hbox{$\left.\vbox to 4em{\vfill}\right\}$}}
	\ht0=0em \dp0=0em 
	~\copy0
$}


\begin{tabular}{|lcl@{}cl@{\hskip 2em}cl|} \hline
\drop{\bfseries Data sources}&\lowarrow&\bfseries \drop{Analysis} & \lowarrow&\bfseries Select results& \lowarrow& \bfseries Submit for\\
\bfseries &&\bfseries & &\bfseries for write up & & \bfseries publication \\ \hline \hline 
&&&&&&\\
Experiments&& Hand calculations &&&&\\
Standard data&& Packages  && Copy \& paste &&\\
Search engines&& SPSS etc & \bigBracket & and edit data && Final paper\\
Literature &&  Graphics packages  && (text, images, graphs, etc)&&\\
Sensors && Specially-written code && into paper &&\\ 
\vdots && \vdots &&&& \\ &&&&&&\\ \hline
\end{tabular}\end{center}}\caption{A simplified schematic of the publication pipeline. For clarity, the pipeline has been linearized; in general, there will be repetitive cyclic iteration and refinement. The RAP and \RAPstar\ approaches encode the normally manual steps in the pipeline processes so that they can be run automatically, and hence reproduce the results that underpin the final paper. The encoded \RAPstar\ algorithms can be shared with other scientists, scrutinized, simplified and optimized, and themselves turned into publishable objects --- they are scientific instruments, just like thermometers or DNA sequencers. Additional schematics are provided in section \ref{RAP-diagrams} in the \supplement.}
\label{fig-pipeline}
\end{table*}

\begin{table*}
{\sf
\newcount \enum
\enum=1
\renewcommand{\theenumi}{%\ifnum \enum<10 \hphantom{0}\fi
\the\enum
\global\advance \enum by 1}
\begin{enumerate}
\item
Peer-review used to ensure the process is reproducible and to identify improvements 
\item
No or minimal manual interference;
for example copy-paste, point-click and drag-drop steps replaced using computer code that can be inspected by others 
\item
Open-source programming languages so that processes do not rely on proprietary software and can be reproduced by others
\item
Version control software, such as Git, to guarantee an audit trail of changes made to code 
\item
Publication of code, whenever possible, on code hosting platforms such as GitHub to improve transparency 
\item
Well-commented code and embedded documentation to ensure the process can be understood and used by others 
\item
Embedding of existing quality assurance practices in code, following guidance set by recognized organizations 
\end{enumerate}
\emph{Adopting RAP principles is not necessarily about incorporating all of the above: implementing just some of these principles will generate valuable improvements.}
}\caption{A minimum standard of RAP, based on the UK Statistics Authority summary \cite{RAPstandard}.}
\label{figure-RAP-summary}
\end{table*}

For clarity, the schematic pipeline in table \ref{fig-pipeline} omits many steps in the creative scientific process. Furthermore, each step is iterated and modified as the research progresses, and, indeed, as referees require revision. The point is that in typical scientific practice each step in the table is largely or entirely manual, typically selecting and copying output from the previous phase, and then pasting and editing the results into the next. The pipeline of data $\rightarrow\cdots\rightarrow$ paper is then iterated by hand as the various components are refined and improved until the authors are happy with the final paper. 

As problems are found in a paper, the data, calculations and code are debugged, refactored, and refined. The process is rarely systematic, and even less likely to be documented --- after all, the atomic steps seem to be innocuous copy and paste actions. The final paper and the ideas it embodies are what matters.

The insight of the reproducible analytic pipelines (RAP) proponents is that every time any step in the pipeline is performed it could have been automated \cite{rap,goldacre,turing-way}. If automated, it could then be repeated reliably --- unlike a manual cut and paste which is error-prone every time it is done. It can be repeated if any experimental data, literature, or other knowledge changes, and the paper's analysis brought up to date with ease. In particular, any other researcher, whether part of the authorship team or a later reader of the paper, can reproduce it reliably if the RAP process is made available. Table~\ref{figure-RAP-summary} provides a brief summary of RAP principles.

For example, if the paper in question is a systematic review, it could be kept current by automatically re-running the programmed atomic actions that it was built with. Indeed, this ability is one of the original motivations of RAP, where Government agencies could easily generate up to date reports on request without having to repeat all the manual work and risk making procedural errors doing so. Furthermore, every time they update a publication, the RAP pipeline itself is reviewed and improved, so the quality of the reproduced work improves --- unlike in a non-RAP process where new errors are generally introduced every time the work is revisited.

RAP not only helps develop reproducible science, and improve the quality of the science as the authors debug and refine their methodology, it also provides a precise audit trail that can be used to protect against fraud, as discussed in \cite{MMR-fraud}, and can in principle be largely automated and perform checks much faster and more efficiently than conventional \emph{post hoc\/} investigations.

RAP embodies Donald Knuth's comment, also quoted after this paper's abstract, 

\begin{quote}\raggedright
{\sf\setbox0=\hbox{``}
\hskip -\wd0\box0
Science is what we understand well enough to explain to a computer'' }
\\ \hfill from the foreword to $A=B$ \cite{a=b}
\end{quote}

The corollary is that if we are doing arbitrary cut and paste that has not been programmed into a computer, then we are not doing good science. Science is in principle an algorithmic process, and therefore, as Knuth says, if we understand well enough what we are doing in science, we can explain it as code, as RAP, for a computer to run and rerun.

\section{The statistics/code analogy}\label{statistics-computation-analogy}
The central role of computational methods in science may be fruitfully compared to statistics, an established scientific tool. 

Poor statistics is much easier to do than good statistics, and there are many examples of science being let down by na\"\i vely planned and poorly implemented statistics. Often scientists do not realize the limitations of their own statistical skills, so careful scientists generally work closely with professional statisticians. 

In good science, all statistics, methods and results are reported very carefully and in precise detail \cite{apa-statistics,Mariusz-reporting,science-statistics-recommendations}, generally following strict journal or disciplinary guidelines. A statistical claim in a paper might be summarized as follows:

\begin{quote}\raggedright
{\sf\setbox0=\hbox{``}
\hskip -\wd0\box0 Random intercept linear mixed model suggesting significant time by intervention-arm interaction effect. 
[\ldots] Bonferroni adjusted estimated mean difference between intervention-arms at 8-weeks 2.52 $( 95\%$ CI $0.78, 4.27, p = 0.0009)$. Between group effect size $d = 0.55$ $(95\%$ CI $0.32,$ $0.77)$.''} \cite{example-stats}
\end{quote}

This wording formally summarizes confidence intervals, $p$ levels, and so on, to present statistical results so the paper's claims can be seen to be complete, easy to interpret, and easy to scrutinize. It is a \emph{lingua franca}. It may look technical, but it is written in the standard and now widely accepted form for summarizing statistics --- it is a clear, rigorous, and readily interpreted way to express uncertainty in results. Moreover, behind any such brief paragraph is a substantial, rigorous, and appropriate statistical analysis.  

Scientists write like this and conferences and journals require it because statistical claims need to be properly accountable and documented in a clear way. The journal \emph{Science}, for example, in its many explicit and quite technical statistics requirements requires 

\begin{quote}\raggedright
{\sf\setbox0=\hbox{``}
\hskip -\wd0\box0 Adjustments made to alpha levels (e.g., Bonferroni correction) or other procedure used to account for multiple testing (e.g., false discovery rate control) should be reported.''} \cite{science-statistics-recommendations}
\end{quote}

Spiegelhalter \cite{Spiegelhalter} says statistical information needs to be accessible, intelligible, assessable, and usable; he also suggests probing questions to help assess statistical quality (see \supplement\ section~\ref{supplementary-Spiegelhalter-section}). Results should not be uncritically accepted just because they are claimed. The skill and effort required to do statistics so it can be communicated clearly and correctly, as above, is not to be taken for granted; in fact, there is widespread concern about the poor quality of statistics in science \cite{paul-hci,pnas-stats}. While it is assumed that statistics should be peer-reviewed, and that review will often lead to improvement, these critical papers show that reviewers and editors are often failing to pick up on poor statistics.

Scientists accept that statistics is a distinct, professional science, itself subject of research and continual improvement. Among other implications of the depth and progress of the field of statistics, undergraduate statistics options for general scientists are insufficient training for rigorous work in science --- their main value, perhaps, is to help scientists to understand the value of collaborating with specialist statisticians. Collaboration with statisticians is particularly important when new types of work are undertaken, where the statistical pitfalls have not already been well-explored.

Except in the most trivial of cases, all numbers and graphs, along with the statistics underlying them, will be generated by computer. Indeed, computers are now very widely used, not just to calculate statistics, but to run the models, do the data sampling and processing, to operate the sensors or surveys that generate the data, and to process it. Many papers now explore the contribution of AI and ML to their fields. The data --- including the databases and bibliographic sources --- and code to analyze it is all stored and manipulated on computers. Computers even help with the word processing and typesetting of the research.

In short, computers, data, and computer code are central to modern science, not just to the explicitly computational sciences. Some AI work is uncovering biases and ethical issues that were previously unrecognized, so computational sciences are not just routinely contributing to existing science but extending its reach and improving its quality.

However, using any code raises many critical questions: formats, backup, cyber-vulnerability, version control, integrity checking (e.g., managing human error), auditing, debugging and testing, and more.  Software code, like statistics, is subject to unintentional bias \cite{Ben,se-bias}.  All these issues are non-trivial concerns requiring technical expertise to manage well. As with statistics, good answers to such ``technical'' issues makes the science that relies on them better.

A common oversight in scientific papers is to present a model, such as a set of differential equations, but omit how that model was reliably transformed into code that generates the results the paper summarizes; if so, the code may have problems that cannot be identified as there is no specification to reference it to, and possibly even no link to the code at all. 

Failure to properly document and explain computer code undermines the scientific value of the models and the results they generate, in the same way as failure to properly articulate statistics undermines the value of any scientific claims. Indeed, as few papers use code that is as well-understood and as well-researched as standard statistical procedures (such as Bonferroni corrections), the scientific problems of poorly planned and reported code are widespread. 

We would not believe a statistical claim that was obtained from some \emph{ad hoc\/} analysis with a new fangled method devised just for one project --- instead, we demand statistics that is recognizable, even traditional, so we are assured we understand what has been done and how reliable results were obtained. 

An interesting overlap with statistical and Software Engineering sloppiness concerns the many papers that  disclose as part of their methodology that they used a particular software package, for example

\begin{quote}\raggedright
{\sf\setbox0=\hbox{``}
\hskip -\wd0\box0 Data analyses were performed using SAS 9.2 (SAS Institute, Cary, North Carolina, USA).''} \cite{sas9.2}
\end{quote}

but without giving any further details. The paper cited above describes one of its multivariate analyses as ``multiple correspondence analysis (MCA) and ascendant hierarchical clustering'' with no specific details --- reproducing the work would be non-trivial since the methodology would have to be reconstructed from scratch, even assuming the data was made available by the authors as the paper gives no details of the data used or how to obtain it. 

The problem is that the common practice of declaring using a named computational system (such as SAS in this case) does not help scrutiny in the least, as such systems can do almost anything. \emph{How\/} those analyses might have been performed is not discussed, and one assumes it follows that the analyses could therefore not have been properly reviewed for basic scientific competence during the publication process. 

A reviewer, if nobody else, needs to actually examine the code used and its documentation to assess whether the analysis presented in the paper is appropriate and sufficiently reliable. Furthermore, if the analysis in this case actually depended on using SAS version 9.2, and not \emph{any\/} general purpose statistical system, then it is problematic because it is not reproducible as it relies on idiosyncrasies in SAS version 9.2. Of course, an author can disclose the idiosyncratic dependencies; while this seems to be an onerous obligation, conversely it is arguable that if an author is unaware of dependencies, then their science relying on them is equally unreliable.

It is recognized that to make critical claims, models need to be run under varying assumptions \cite{whitty}, yet somehow it is easy to overlook that the code that implements those models also needs to be carefully tested under varying assumptions to uncover and fix bugs and biases, as well as to uncover unknown dependencies. Indeed, the code may be poorly written (as this paper shows it is), so the results derived from the code simply may not be reliable. 

\label{section-vicious-cycles}
In normal scientific reporting (outside of teaching and assessing science) details of methodology are routinely glossed. A chemist does not say they cleaned their glassware. One might argue, then, that scientists need not discuss their code in detail because they know how to program and their code is correct. This argument is mistaken. Code is rarely considered a valuable part of the science to which it contributes, which creates a vicious cycle of ignoring code, leading to ignoring the critical --- and non-trivial --- role of correct code in science. 

%\color{blue}
\input generated/over-fitting-code-section.tex
%\color{black}

\section{The conventional role of code}

Models map theory and parameters to describe phenomena, typically to make predictions, or to test and refine the theory supporting the models. With the possible exception of theoretical research, all but the simplest models require computers to use; indeed even theoretical mathematics is now routinely performed by computer.

Whereas the mathematical form of a model may be concise and readily explained, even a basic computational representation of a model can easily run to thousands of lines of code, and its parameters --- its data --- may also be extensive. The chance that a thousand lines of code is error free is negligible, and therefore good practice demands that checks and constraints should be applied to improve its reliability. How to do this is the concern of Software Engineering. 

While scientific research may rely on relatively easily-scrutinized mathematical models, or models that seem in principle easy to mathematize, the models that are run on computers to obtain the results published are sometimes not disclosed, and even when they are they are long, complex, inscrutable and (as our survey shows) lack adequate documentation. Therefore the models are very likely to be unreliable \emph{in principle}. 

If code is not well-documented, this is not only a problem for reviewers and scientists reading the research to understand the intention of the code, but it also causes problems for the original researchers themselves: how can they understand their historical thinking well enough (say, just a few weeks or months later) to maintain it correctly if it has not been clearly documented? As a scientist pursues a research career building on their previous work, how can they be certain their work is reliable, and not merely converging to their  prejudices? Without proper documentation, including a reasoned case to assure that the approach taken is appropriate \cite{assurance-case}, how do researchers, let alone reviewers, know exactly what they are doing?

Without substantial documentation it is impossible to scrutinize code properly. Consider just the single line ``\texttt{y = k*exp(x)}'' where there can be \emph{no\/} concept of its correctness \emph{unless\/} there is also an explicitly stated relation between the code and the mathematical specifications. What does it mean? What does \texttt{k} mean --- is it a basic constant or the result of some previous complex calculation? Does the code mean what was intended? What are the assumptions on \texttt{k}, \texttt{x}, and \texttt{y}, and do they hold invariantly? Moreover, as code generally consists of thousands of such lines, with numerous inter-dependencies, plus calling on many complex libraries of support code, it is inevitable that the \emph{collective\/} meaning will be unknown. A good programer would (in the example here) at least check that \texttt{k} and \texttt{x} are in range and that \texttt{k*exp} was behaving as expected (e.g., in case of under- or overflow).

Without explicit links to the relevant models (typically mathematics), it is impossible to reason whether any code is correct, and in turn it is impossible to scientifically scrutinize results obtained from using the code. Not providing code and documentation, providing partial code, or providing code without the associated reasoning is analogous to claiming ``statistical results are significant'' without any discussion of the relevant methods and statistical details that justify making such a claim. If such an unjustified pseudo-statistical claim was made in a scientific paper, a reviewer would be justified in asking whether a competent experiment had even been performed. It would be generous to ask the author to provide the missing details so the paper could be better reviewed on resubmission. 

Some authors assert that the purpose of code is to provide insight into models, rather than precise (generally numerical) analyses summarizing data or properties of the data \cite{assessing-quality}. In reality, if code is inadequate, any so-called ``insights'' will be flawed, and flawed in unknown ways. Indeed, none of the papers sampled (described in \supplement\ section~\ref{survey-section}) claimed their papers were using code for insight; all papers claimed, explicitly or implicitly, that their code outputs were integral to their peer-reviewed results.

Clearly, like statistics, coding can be done poorly and reported poorly, or it can be done well and reported well --- and any mix between the extremes. The question is whether it matters, \emph{when\/} it matters, and, if so, when it does, \emph{what\/} can be done to \emph{appropriately\/} help improve the quality of code (and discussions about the code) in scientific work?

\subsection{The deceptive simplicity of code}\label{deceptive-simplicity-of-code}
It is a misconception that programming is easy and even children can do it \cite{fixit}. More correctly, toy programming is easy, but mature programming is very difficult.

An analogy helps justify this key point. Building houses is very easy --- indeed, many of us have built toy Lego houses. Obviously, though, a Lego house is not a \emph{real\/} house. It is not large enough or strong enough for safe human habitation! This point is obvious because we can see Lego houses, and everyone is familiar with building-block play. Its real-world engineering limitations are too obvious to need stating. 

In contrast to Lego, computer programs are generally invisible, and therefore the engineering problems within them are also made invisible. The ``programming is easy'' clich\' e is deceptive --- programming appears easy \emph{because\/} professional standards of building software are ignored, because people cannot see the reasons why they are needed, and because --- like Lego --- toy programs can look inspiring but be unreliable, difficult to use, even dangerous. 

Saying programming is easy is like appreciating a child's Lego building because we are not worried about subsidence, load bearing, electric shock, fire risks, water ingress, or even planning regulations. These are professional engineering issues that Lego builders ignore. Certainly, even real building is much easier and faster when the technical details are ignored, as anyone who has experienced a cowboy builder can attest.

Unlike building houses (the Code of Hammurabi dates to around 1755{\sc bc}), programming is a new discipline, and the problems of poor programming are not widely appreciated or embedded in our culture. Professional standards, even when they exist, are not enforced. 

Problems for the reliability of science arise when doodling and tweaking software drifts into claiming scientific results that do not have reliable engineering processes or structures underpinning them (let alone the properly developed and documented accessible code) to justify them. 

In many countries, there are laws that require all but the very simplest building structures to be formally approved from plans and inspected as they are built, but who writes plans for software, who inspects scientific models while they are being coded? Yet the consequences of building a shoddy garage have negligible impact compared to the consequences of writing poor code that informs national public health policies or climate change interventions.

\subsection{The low status of coding}
Since programming appears to be so easy, developing code has a correspondingly low status in scientific practice (and more widely). Developers of code are rarely acknowledged in scientific papers. The implicit reasoning is: if programming is easy, then its intellectual contribution to science is negligible, so it is not even worth citing it or acknowledging the contributors to it. Because it is apparently easy, there is no need to work hard to make it correct. Because of the ease of over-fitting (section \ref{over-fit}), code ``works well'' with little skill or effort. While such mistaken views prevail, the vicious cycle is that the low status means software development is casualized, which reinforces the low status.

Almost all scientific papers \emph{routinely\/} describe their experimental method, their data handling, and provide an overview of their analytic (usually statistical) methods. If they are theoretical papers, they will describe their mathematical models and data that is used to run or test their models. Outside of pure computer science, scientific papers are almost entirely silent on the code they rely on and how it was developed --- in particular, how the code might have been protected from bugs, analogously to how appropriate experimental methods were used avoid or control for experimental error. 

%An example is the documentation at the code repository \cite{BayesWaveCode}, which requires three high-profile \emph{{Physics Review D}} papers \cite{BayesWave1,BayesWave2,BayesWave3} be cited to acknowledge any use, but only one of them, \cite{BayesWave3}, mentions the code, and even then it fails to acknowledge the contributions of its 10 programmers (despite 6 being co-authors). The paper does cite the repository URL, but unlike the papers it cites conventionally, it omits all the code's authors. Acknowledgement does not seem to matter, even to the people who did the work.

In reaction to this vicious cycle, there is a growing movement to cite code correctly \cite{cite-code,TOP}, because  code \emph{is\/} important, particularly for reproduction, testing and extension of any scientific work. Curiously, the importance of \emph{correct\/} code is rarely emphasized.

Few journals editorial policies recognize that data and code are theoretically and in practice indistinguishable (see \supplement). Given that data and code are equivalent and interchangeable, it follows that publishing policies on data handling should also apply at least as strictly to code.

\subsection{The critical role of code is often ignored}\label{central-role-of-code}\label{PRISMA-statement}
Because statistics, like code, is so readily susceptible to uncontrolled bias and error, there are many protocols and journal policies that enforce best practice, for example journals often require adherence to PRISMA (Preferred Reporting Items for Systematic Reviews and Meta-Analyses) \cite{prisma} for any paper performing a systematic review of the literature. Yet PRISMA, like many such policies, ignores the critical role of code, and ignores the Software Engineering principles that assure code that research relies on is reliable and reliably reported. 

PRISMA ``was designed to help systematic reviewers transparently report why the review was done, what the authors did, and what they found,'' which is reasonable enough. PRISMA covers the review process carefully. For example, the authors should report the number of papers they included in their review. Perhaps $N=2000$. This number is then written into their draft paper, perhaps in several places. As the authors read and revise their paper, submit it, and respond to peer-reviewers, it is very likely that the number of papers in the survey changes, or other numbers and details may change. 

The authors now have a maintenance problem: where are the numbers that have changed, and what should they be changed to? Doing a search-and-replace, whether automated or by hand, is fraught with difficulties. What happens if $2000$ is used for some other purposes as well? What happens if some of the $2000$ values are written as $2,000$ or as $2000.0$? What happens if some $2000$ are year dates and are changed incorrectly? Then there are the Human Factors: slips and errors will happen in this process anyway \cite{engima}. Typos, slips during cut-and-paste, and other errors are common. Similar iterative revision cycles happen with any paper, not just with systematic reviews. PRISMA rightly deals with a wide range of methodological problems, but ignores the methodological problems (such as the example earlier in this paragraph) that using code na\"\i vely raises. 

The irony is that PRISMA says nothing about how to ensure the results of a survey are correctly and reliably presented in a paper, despite this being one of PRISMA's explicit motivations. Such rule schemes reinforce the fallacy that code is trivial and unimportant. PRISMA consciously warns about  methodological issues, but ignores that poorly managed code raises comparable methodological issues that undermine the validity of the research it supports.

\subsection{Bugs, code and programming}\label{knowledge}
Critiques of data and model assumptions are increasingly common \cite{critiques,diagnosis-reviews} but program code is rarely mentioned. Yet data and program are formally equivalent (see \supplement, section \ref{on-code-data-publication}). Program code has as great an effect on results as the data; in fact, without code, the data would be uninterpreted and probably useless. Code, however, is harder to scrutinize, which means that errors in code have subtle, often unnoticed effects on results.

Almost all code contains ``magic numbers'' --- that is, data masquerading as code. This common practice ensures that published data is very rarely all of the data because it omits the magic numbers embedded in the code. Such issues emphasize the need for repositories to require the inclusion of code so all data, including that embedded in the code, is actually available. 

Bugs can be understood as discrepancies between what code ought to do and what it actually does. Many bugs cause erroneous results, but bugs may be ``fail safe'' by causing a program to crash so no incorrect result can be delivered. Contracts and assertions are essential defensive programming technique that block compilation or block execution with incorrect results; they turn bugs into safe termination, or, better, failure to compile. None of the science surveyed in this paper includes any such basic techniques. 

Random numbers are widely used in computational science (and in many of the papers surveyed), for simulation or for randomizing experiments. Misuse of random numbers (e.g., using standard libraries without testing them) is a very common cause of bugs nai\"\i ve bugs \cite{knuth}.

If code is not documented it cannot be clear what it is intended to do, so it is not possible to detect and eliminate bugs. Indeed, even with good documentation, \emph{intentional bugs\/} will remain, that is, code that correctly implements the wrong things \cite{essence-of-software,fixit} --- they are bugs that were intended but were ideas based on mistaken ideas (students and inexperienced programmers make intentional bugs all the time). For instance, in numerical modeling, using an inappropriate method can introduce errors that are not ``bugs'' in the narrow sense of incorrectly implementing what was wanted, but are bugs in the wider sense of producing incorrect results (e.g., ill-conditioning); that is, what was intended was~wrong. 

\subsection{Long-term problems of unreliable code}
Scientists explore and extend the boundaries of rigorous knowledge. Put briefly, the purpose of scientific experiments is to vary details to either test and specify the boundaries of theories, or to discover new phenomena that then lead to theory revision. 

If poor code, or poorly documented code, is made available with scientific papers, the code is a natural place to start replicating and varying experimental conditions,  including both data or code. However, if the starting point is not accurately known, whether due to bugs, obscure code, or because of poor documentation, then experimental variations will have an unknown effect. Theory will then be driven by artifacts of the code, not genuine phenomena. 

In section \ref{section-pandemic-modeling}, below, an example is documented of a research code development process of at least 15 years' duration where the code was admitted to be completely undocumented, leaving details in just one author's head. None of the various related papers describe any controls over the drift of the science, or how independent researchers building on it might have been able to build with confidence rather than merely reproducing the same errors. 

Since the code in question was substantial and non-trivial, it is very unlikely that any constructive reproduction occurred outside the original laboratory and mindset; indeed, section \ref{reproducibility-concerns} describes how ``reproduction'' became trivialized because of community pressure to confirm the insights of this particular research. 

Trying to constructively refute aspects of this research in the Popperian sense \cite{popper-conjectures-refutations} would have been impossible. For example, had the relevant papers published critical code invariants then scientists building on the research could have explored whether those invariants remained valid and, if so, under what assumptions. In fact, invariants are the theories of code, and deserve as high a prominence in published  science as the domain theories the code itself is supporting investigating. 

%Since the symmetries in the scientific domains should be reflected in corresponding symmetries in well-designed code, this raises new possibilities of Formal Methods approaches (e.g., using automatic theorem provers) supporting advances in general science.

\section{State of the art}% \\ in computational science}

\subsection{Case study: Pandemic modeling}
\label{section-pandemic-modeling}
A review of epidemic modeling \cite{science-review} says, ``we use the words `computational modeling' loosely,'' and then, curiously, the review discusses exclusively mathematical modeling, implying that for the authors, and for the peer-reviewers, there is no role for code or computation as such. It appears that the new insights, advances, rigor, and problems that computers bring to research were not considered relevant. 

A systematic review \cite{diagnosis-reviews} of published COVID models for individual diagnosis and prognosis in clinical care, including apps and online tools, noted the common failure to follow standard TRIPOD guidelines \cite{tripod}. The review \cite{diagnosis-reviews} itself ignored the mapping from models to their implementation, yet if code is unreliable, the model \emph{cannot\/} be reliably used, and cannot be reliably interpreted regardless of whether TRIPOD guidelines are followed. Indeed, TRIPOD guidelines ignore code completely. 

It should be noted that flowcharts, which the review \cite{diagnosis-reviews} did consider, are programs intended for human use. Flowcharts, too, should be designed as carefully as code, for exactly the same reasons. 

A high-profile 2020 COVID-19 model \cite{nature-summary,ICmodel} uses a modified 2005 computer program \cite{avianFluModel,originalICmodel} originally developed for modeling H5N1 in Thailand, when it did not model air travel or other factors required for later western COVID-19 modeling. The 2020 model forms part of a series of papers \cite{ICmodel,avianFluModel,originalICmodel} none of which provide details of their code. 

A co-author disclosed \cite{tweet} that the code was thousands of lines long and was undocumented C code. As Ferguson, the original code author, noted in an interview, 

\begin{quote}\raggedright
{\sf\setbox0=\hbox{``}
\hskip -\wd0\box0 
For me the code is not a mess, but it's all in my head, completely undocumented. Nobody would be able to use it~\ldots
''} \cite{ferguson-interview}\end{quote}

This comment was made by a respected, influential world-leading scientist, with many peer-reviewed publications, and a respectable $h$-index\footnote{$h$-index: the largest value of $h$ such that at least $h$ papers by the author have each been cited at least $h$ times. The figure cited for Ferguson was obtained from Google Scholar on 20 January 2022. (Typical $h$ values vary by discipline.)} of 93. Ferguson should be well aware of the standards of coding used in his own field. Arguably, then, this comment quoted above is representative of the culture of the field.

Ferguson's admission is tantamount to saying that the published scientific findings are and need not be reproducible.\footnote{A constructive discussion of Software Engineering approaches to reproducibility can be found in \cite{basic-reproducibilty}.}  

Lack of reproducibility is problematic, especially as the code would have required many non-trivial modifications to update it for COVID-19 with its different assumptions; moreover, the code would have had to have been updated very rapidly in response to the urgent COVID-19 crisis. 

If Ferguson's C code had been made available for review, the reviewers would not have known how to evaluate it without the relevant documentation. It is, in fact, hard to imagine how a large undocumented program could have been repeatedly modified and repurposed over fifteen years without becoming incoherent. 

If code is undocumented, there would be an understandable temptation to modify it arbitrarily to get desired results; worse, without documentation and proper commenting, it is methodologically impossible to distinguish legitimate attempts at debugging from merely fudging the results. In contrast, if code is properly documented, the documentation defines the original intentions (including formally using mathematics to do so), and therefore any modifications will need to be justified and explained --- or the theory revised.

The programming language C which was used \cite{tweet} is, like many popular programming languages, not a dependable language; to develop reliable code in C requires professional tools and skills. Some of the code was written in a na\"\i ve style (e.g., writing \texttt{*(a + i)} instead of \texttt{a[i]}, and with obscure numerical goto statements like \texttt{if(l == 0) goto S150}), and with C code that was translated simplistically from FORTRAN and Pascal code, from references dating from the 1970s and 1980s \citeeg{forsyth,Normal-Distributions}.

Moreover, C code is not portable, which limits making it available for other scientists to use reliably: C notoriously gets different results with different compliers, libraries, and hardware. In fact, in any area where reliable programming is required in a C-like language, a special dialect such as MISRA C should be used: MISRA C manages the serious design flaws of C that otherwise make it too unreliable \cite{misra}. Alternatively, a high integrity programming language, unrelated to C, such as SPARK Ada \cite{spark}, or modern languages (many related to the ``ML family'') like OCaml, F*, Haskell \cite{haskell}. These languages have learning curves; however their key benefit is that correct programs are far more likely and are \emph{much\/} faster to write. (The \supplement\ discusses these issues further.) 

Ferguson, author of the code, says of the criticisms of his code, 

\begin{quote}\raggedright
{\sf\setbox0=\hbox{``}
\hskip -\wd0\box0 
However, none of the criticisms of the code affects the mathematics or science of the simulation''} \cite{thumbs-up}
\end{quote}

This claim is implausible. 

The original work on theoretical epidemiology may be fine if it does not use any of his code, but if the science is not supported by code that correctly implements the models, then the program's output cannot be relied on without independent evidence. Over the fifteen plus years the code was in development the science it informs will have developed too, as will the relevant data; it is not clear how they will have remained in alignment.

\label{discuss-covid-sim}
Typically, models will be developed iteratively as their results are improved to better fit a scientist's goals --- but this, especially when it is done by tinkering, as here --- risks making the code arbitrarily fit the goals (that is, over-fitting; see section \ref{over-fit}), rather than to objectively elucidate the science. 

In fact, the Ferguson code, \texttt{covid-sim}, is very large at \covidsimkLOC\ kLOC (thousands of lines of code),\footnote{Ferguson's \texttt{covid-sim} system is composed of \covidsimfiles\ files, and uses \covidsimdata\ of data. It is now rewritten from C into C++ with Python, R, sh, YAML/JSON, etc. For more details, see \supplement.} so it is implausible that the ``mathematics or science'' has been correctly implemented in it. Ferguson's \emph{reported\/} science unlikely to be reliable. 

\subsection{Concerns with reproducibility}
\label{reproducibility-concerns}
Getting science right, which now, in turn, depends on correct code, is a normal requirement of \emph{reproducibility}.

The code in \cite{nature-summary,ICmodel} has been ``reproduced,'' as reported in \emph{Nature\/} \cite{codecheck,thumbs-up}, but this so-called reproduction merely confirmed that the code could be run again and produced comparable results (compared using an Excel spreadsheet). This weak test would pass provided the runs gave the same invalid answers --- it is not usefully a stronger test than just checking that the code compiles.

Running code to obtain results claimed in a paper is a very weak test, and anyway one that should have been checked automatically during paper preparation and submission. If you do not know what you are reproducing, as is the case in this \emph{Nature\/} paper, there is little scientific value in doing so.

Unfortunately, the terms reproducibility, replicability, and repeatability, have similar meanings in English but have been used in different specific technical ways by different authors. In \cite{codecheck,thumbs-up} the reproduction amounted to just re-running the original code. It is certainly essential to establish that a paper's code can be run, as non-working code cannot support any claims in a paper; if the original code runs this confirms a basic level of access for the wider scientific community. But a more realistic criterion than basic reproduction in this sense is whether an \emph{independently\/} developed model developed from the same paper(s) specifications produces equivalent results (called $N$-version programming, a standard Software Engineering practice \cite{NVP}) like public health surely requires as, indeed, Ferguson's own influenza paper \cite{nvp-ferguson} argues. However, much stronger scrutiny of code than ``reproduction'' is required to answer essential questions including:

\newcount \enum
\enum=1
\renewcommand{\theenumi}{\ifnum \enum<10 \hphantom{0}\fi
\the\enum
\global\advance \enum by 1}
\begin{enumerate}\raggedright
\item Is the code valid: does it do what the paper claims?\footnote{Of course, the underlying science may be wrong to, so it is useful to distinguish \emph{internal validity\/} and \emph{external validity}. Internal validity occurs if the code does what the paper claims; external validity occurs if the code represents correct science (which the paper may have interpreted incorrectly.}
\item Do other scientists, including reviewers and the authors, understand the code?
\item Does the code implement the methods described in the paper?
\item Has the code been over-fitted or tweaked to support specific claims in the paper?'
\item Is there a definitive version of code?
\item Is the code controlled and signed?
\item What limitations does the code have?
\item Was the code developed to any standard, and does it comply to that standard?
\item How does the code protect against data, coding, and human error?
\item Was the code tested adequately?
\item Does the code depend on arbitrary parameters, data, or code to over-fit to obtain the published results?
\item Is the code documented adequately, so we know what it is trying to do, and how?
\item \ldots\ and so forth --- questions are numbered for reference. 
\end{enumerate}

All such questions also apply to specifications, documentation, assurance cases, test procedures, and other essential documents, not just to code. In turn, the levels of scrutiny demanded should be guided by explicit claims in the paper \cite{essence-of-software} --- for example, a pilot study requires weaker assurance than code that is developed concerning nuclear power, driverless vehicles, public health, etc. 

The questions in the list above are certainly hard to answer for all but the briefest code, but corresponding levels of quality assurance are demanded for other methodologies \cite{notebooks,popper-conjectures-refutations,tripod,prisma,nih-policy,nih-nature}, such as data preparation and statistics to support claims in peer-reviewed science. 

Because of the recognized importance of the Ferguson paper, a project started to document its code  \cite{refactoring}.\footnote{The system is now open-source, available at \url{github.com/mrc-ide/covid-sim} version (19 July 2021).} Documenting code in hindsight, even if done rigorously, may describe what it does, \emph{including\/} its bugs, but it is unlikely to explain what it was originally intended to have done. As the code is documented, bugs will be found, which will then be fixed (refactoring), and so the belatedly-documented code will not be the code that was used in the published models; it will be different. 

It is well-known that documenting code helps improve it, so it is surprising to find an undocumented model being used in the first place, since so many years' opportunity to improve the code have been lost. The revised code has now been published, and it too has been heavily criticized \citeeg{bad-code}, supporting the concerns expressed in the present paper.

Some papers \citeeg{pseudo} publish models in pseudo-code, a simplified form of programming. Pseudo-code looks deceptively like real code that might be copied to try to reproduce it, but pseudo-code introduces invisible and unknown simplifications. Pseudo-code, properly used, can give a helpful impression of the overall approach of an algorithm, certainly, but pseudo-code alone is not a surrogate for code: using it instead of making actual code available is worse than not publishing code at all (see \cite{chinese}). Pseudo-code is too vague to help anyone scrutinize a model; moreover, pseudo-code may mask over-fitting in code used that is not explicit in the pseudo-code. 

An extensive criticism of pseudo-code, and discussion of tools for reliable publication of code can be found elsewhere \cite{relit}. 

The \supplement\ provides further discussion of reproducibility.

\subsection{Beyond pandemic modeling}
\label{section-science-beyond-pandemic-modeling}

Epidemiology has a high profile because of the COVID pandemic, but the problems of unreliable code are not limited to COVID-19 modeling papers, which, understandably, were perhaps rushed into publication. But other examples that were not rushed include a 2009 paper reporting a model of H5N1 pandemic mitigation strategies \cite{flu-model}, which provides no details of its code. Its supplementary material, which might have provided code, no longer exists.

There are many other areas of computational science that are equally if not more critical, and many will have longer-lasting impact. Climate change modeling is one such example that will have an impact long beyond the COVID pandemic.

A short 2022 summary of typical problems of Software Engineering impacting science appears in \emph{Nature\/} \cite{nature-review}, describing diverse and sometimes persistent problems encountered during research in cognitive neuroscience, psychology, chemistry, nuclear magnetic resonance, mechanical and aerospace engineering, genomics, oceanography, and in migration. The paper \cite{nature-review}  makes some misleading comments about the simplicity of Software Engineering, e.g., ``If code cannot be bug-free, it can at least be developed so that any bugs are relatively easy to find.''

Guest and Martin promote the use of computational modeling \cite{psychological-modeling}, arguing that through writing code, one debugs scientific thinking. Psychology, their focus, has an interesting relationship with software, as computational models are often used to model cognition and to compare results with human (or animal) experiments \cite{psychological-modeling}. In this field, the computation does not just generate results, but is used to explicitly explore the assumptions and structures of the scientific frameworks from which the models are derived. Computational models can be used to perform experiments that would be unethical on live participants, for instance involving lesioning (damaging) artificial neural networks. It should be noted that such use of cognitive models is controversial --- on the one hand, the software allows experiments to be (apparently) precisely specified and reproduced, but on the other hand in their quest for psychological realism the models themselves have become very complex and it is no longer clear what the science is precisely!

For instance, ACT-R, one widely-used theory for simulating and understanding human cognition, has been under development since 1973, and is now a 120 kLOC Common LISP and Python system \cite{actr}. Furthermore, any paper using ACT-R would require additional code on top of the basic ACT-R framework.

The psychology paper \cite{psychological-modeling} presents an example computational model from scratch to illustrate a framework of computational science. In fact their example model has no psychological content: a simple numerical test is performed, but the psychology of why the result is counterintuitive --- the psychological content --- is not modeled. Be that as it may, they develop a mathematical specification and discuss a short Python program they claim implements it. 

The Python code is presented without derivation; Software Engineering is ignored. The program listed in the paper certainly runs without obvious problems (ignoring some typographical errors due to the journal's publishers), but ironically the Python does \emph{not\/} implement the mathematical specification explicitly provided for it, thus undermining the argument of the~paper. 

One might argue the bug is trivial (the program prints \texttt{False} when it should print \texttt{b}), but to dismiss such a bug would be comparable to dismissing a statistical error that says $p=\mbox{\tt False}$ which would be nonsense --- if a program printed that, one would be justified in suspecting the quality of the entire program and its analyses. Inadvertently, it would seem, then, that the paper shows that just writing code does not help debug scientific thinking: instead, code must first be derived in a rigorous way and actually be correct. Otherwise, code based on inadequate Software Engineering will introduce errors into scientific thinking.

Code generally for any field of scientific modeling needs to be carefully documented and explained because all code has tacit assumptions, bugs and cybersecurity vulnerabilities \cite{Ben,nature-review,se-bias} that, if not articulated \emph{and properly managed}, can affect results in unknown ways that may undermine any claims. People reading the code will not know how to obtain results because they do not know exactly what was intended in the first place. The problem is analogous to the problem of failing to elaborate statistical claims properly: failure to do so suggests that the claims may have unknown limitations or flaws.

Even good quality code has, on average, a defect every 100 lines --- and {such a low} rate is only achieved by experienced industrial software developers \cite{ourReview}. World-class software can attain maybe 1 bug per 1,000 lines of code. Code developed for experimental research purposes will have higher rates of bugs than professional industrial software, because the code is less well-defined and evolves as the researchers gain new ``insights'' into their ideas, unable to distinguish genuine insights from artifacts of bugs. In addition, and perhaps more widely recognized, code --- especially but not exclusively mathematical code --- is subject to numerical errors \cite{hamming}. It is therefore inevitable that typical modeling code has many bugs (reference \cite{NVP} is a slightly-dated but very insightful discussion). Such bugs undermine confidence in model results. 

Only if there is access to the \emph{actual\/} code and data (in the specific version that was used for preparing the paper) does anyone know what the researchers have done and whether that corresponds closely to what they are reporting.

Some COVID-19 papers \citeeg{unfinished} make unfinished, incomplete code available. While some \citeeg{unfinished,lancet-unfinished} make what they call ``documented'' code available, they provide no more than superficial comments. This is \emph{not\/} documentation as properly understood. Such comments do not explain code, explain contracts, nor explain algorithms. Contracts, for instance, originated in work in the 1960s \cite{hoare}, and are now well-established practice in reliable programming.

Even if a computer can run it, badly-written code (as found in \emph{all\/} the research reviewed in the present paper, and indeed in computer science research, e.g., \cite{machine-learning-reproducibility}) is inscrutable. Only if there is access to \emph{adequate\/} documentation can anyone know what the researchers \emph{intended\/} to do. Without all three (code, data, adequate documentation), there are dangers that a paper simplifies or exaggerates the results reported, and that omissions, bugs and errors in the code or data, generally unnoticed by the paper's authors and reviewers, will have affected the results they report \cite{relit}. 

\label{section-discussion}
\subsection{The narrow emphasis on data}
\label{critique-fair}
Data has been at the center of science, certainly since the earliest days of astronomy collecting planetary and other information. Today it is widely recognized that lack of accessible and usable data that has already been collected limits the progress of science. Low quality data and poor access to data causes reproducibility problems, an increasingly recognized problem --- in 2015 it was estimated that \$28 billion a year is spent on preclinical research that is not reproducible \cite{preclinical-reproducibility}.

Curating data is taken seriously as a part of normal science and peer-reviewed publication. Journal policies widely require appropriate discussion of data, much as they require appropriate discussion of statistics. Journals often require archiving data in standard formats so it can be accessed for reproduction in further scientific work. 

There are many current activities to proceduralize and standardize the more effective curation and use of data, such as the FAIR principles (Findable, Accessible, Interoperable and Reusable) for scientific data management and stewardship \cite{fair,fair-principles}, and in the development of journal and national funder policies. For example, the 2022 update to the US National Institutes of Health data policies \cite{nih-policy} is described as a ``seismic mandate'' by \emph{Nature\/} \cite{nih-nature} in its attempt to improve reproducibility and open science yet they ignored code.

These cost estimates and initiatives under-play the role of code as a critical component despite its becoming the new laboratory for almost all science. The role of code specifically in modeling is discussed throughout this paper; without bespoke code, proposed models (unless intended to be abstract) cannot make a quantifiable contribution to the literature. Code has additional problems of versions and compatibility beyond those of data, for example suitable compilers to run old code may no longer be available, and  programming systems may produce different results when used on different computers. 

In general, without proper management of code --- for example to record, detect and report version control differences --- sharing code may even be counter-productive.\footnote{The data and code shared with the present paper includes cryptographic checksums; if somebody reproducing the work described here does not obtain the same checksums at least when they start their work, then there are problems that need investigation before relying on the reproducibility of the data.}

Using structured repositories that provide suggestions for and which encourage good practice (such as Dryad and GitHub), and requiring their use, would be a lever to improve the quality and value of code and documentation in published papers. The evidence (see \supplement) suggests that, generally, some but rarely all develop code that is uploaded to a repository just before submitting the paper in order to ``go through the motions.'' In the surveyed papers there is no evidence (before, during, or after the date of the survey sample) that any published code was prepared or maintained \emph{using\/} repositories. This is consistent with the finished code being uploaded to a repository just for the purposes of satisfying publishing requirements, but not using one earlier probably because they did not understand the benefits of doing so. 

Using a standard repository for lodging a paper's supporting code would help other scientists access the code easily, but not using the repository for developing and maintaining the code means the author of the paper misses out on many helpful features of repositories, such as version control, open-source development and review, actions and other approaches for automating development, sharing workload, and so on.

In summary, there is a lop-sided emphasis on data in science. In fact, data is useless without code, and code must be used to analyze data and can manipulate it in arbitrary ways --- some ways, like renormalization and other transformations, deliberately, and some accidentally caused by bugs and unwittingly misunderstanding programming techniques. Often code is used to extrapolate data, so the code itself effectively generates more data, or the code eliminates outliers so it effectively deletes data. Data is routinely formatted in standard ways (else code could not process it), but code is generally very architecture- and version-specific, so --- unless properly managed --- code goes obsolete faster than data. And so on. In short: the integrity of code and its availability to scrutiny is in fact both harder and more important than the usual requirements put on data.

\section{Extending RAP to \RAPstar}\label{RAPstar-section}
Code is usually seen as an independent set of files that are run to generate results, typically to be copied into a paper. But we can go further: a paper can itself embed code or become code \citeeg{relit,Executable-Paper}. This supports the generalization of RAP to form \RAPstarp. Essentially, \RAPstar\ is the recognition that computing is not just about programming computers (which results in RAP) but is about applying computational thinking \cite{computational-thinking-wing,computational-thinking} that supports and constructively analyzes any process, and in particular human and scientific processes that may not normally be computer-based at all.

Once processes in the pipeline are automated, this means that there is code to run the steps again. Once there is code, it can be managed in a version control system. A version control system then provides an audit trail for free, as well as many advantages such as being able to backtrack to an earlier version to review earlier edits. Importantly, code can also perform sanity checks on the process --- a very simple example is automatic bibliography systems that check that journal names, DOIs are correct, and so forth. (They also allow the bibliographic data to be pooled and curated with other scientists, which improves its scope and quality.)

GitHub is a tool that provides \emph{actions\/}, which are named specifications to run RAP's analytic pipelines, or workflows in GitHub's terms. GitHub happens to specify actions in the language YAML, which, being a textual notation, in turn means that all the helpful features of GitHub --- open-source, version control, and so on --- can be applied to the pipelined processes as well. Research pipelines can thus be made explicit, documented, shared and improved with open collaboration.

Almost the entire scientific process can be automated (even with its interaction with the world automated by sensors and robots). There are many ways to do this; for example, \emph{Mathematica\/} makes the analysis of the data and the calculations and the paper ``the same thing'' in its integrated notebook user interface. The many alternatives include R Markdown, an approach based in the open mathematical system R \cite{RMarkdown}; a system, Lepton \cite{lepton}, which allows a \LaTeX\ document to execute and include arbitrary code, and language-independent notebook systems like Jupyter; and so on. There are also variations of literate programming, notably relit \cite{relit}, that enable papers themselves to generate the code they rely on and may describe, which therefore ensures the code run to support papers is exactly the code the author ran to support the claims. 

In all such systems, running the computational paper creates the publication. Indeed, every time the paper is run, the authors are likely to check the results after they have been re-computed, so the RAP process actively helps reduce errors. 

Here is the insight: the code, just like the paper is text, so the code \emph{itself\/} can fully form part of the pipeline, made reproducible and benefit from all the RAP benefits. To date, this critical point has been overlooked. Since using RAP to improve code, rather than just the pure scientific pipeline, has not been mentioned previously, we call it \RAPstar\ to make it clear this is a new and important generalization. \RAPstar\ helps improve code quality, for the same reasons RAP improves the quality of the paper and the science. 

Software engineers have many tools for automatic code development (such as Unix's \texttt{make}) but the idea that these tools can be used to integrate and help automate code authoring as well as its documentation \emph{and\/} paper authoring is radical. Amongst other things, it means that the \emph{entire\/} research and development process of the paper \emph{including\/} all its underlying code can be reproduced and reused by others. The present paper is a modest example of \RAPstarp; more details are given in the \supplement. 

Note that as \RAPstar\ objectifies how science is done to a standard sufficient to enable a computer to run it. \RAPstar\ enables all of the methodologies of Software Engineering to be brought to bear \emph{on the science itself}. \RAPstar\ means the normally tacit, manual, and undocumented processes of science become explicit code. Code can then be scrutinized, optimized, and ensured correct by standard Software Engineering practice --- thus \RAPstar\ does not just automate science for reproduction, it makes the automation explicit so the doing of science itself can be reasoned about --- not just mentally but supported by sophisticated tools, such as theorem proving and AI\@. Science will be improved.

\section{The paper as a scientific laboratory}\label{paper-as-lab}
The conventional view of science is that experiments are done \emph{then\/} written up. 
%RAP and \RAPstar\ support and enhance this process with computational tools; in particular, the archival results of research are not just the reports and publications, but include the executable processes --- the pipelines in RAP terms --- of the research. 
However, it is more productive to think of the paper itself as an active laboratory, not just as a record of finished work. Viewing the paper itself as a laboratory encourages authors to copy and adopt laboratory best practice (such as keeping records, as RAP and \RAPstar\ suggest) into the preparation of the paper itself; it also encourages authors to see writing as a scientifically --- not just expressive --- creative act, and not just as the final summary of a period of scientific creativity that is not written down. In short, seen as a laboratory, writing the paper is no longer a reactive write-up of finished work, but it is an active part of doing the science itself. Writing a paper thus explores the space of scientific possibility as constructively as working in the field or on a bench: the paper is a laboratory.

With a computational paper, authors can literally experiment \emph{in\/} the paper, exploring the effectiveness of ideas and explanations. Furthermore, they can experiment with hypotheses: for example, they can make a clear claim that at that point in the lifecycle of the paper is but a wish rather than an established fact --- sketching a direction they plan to go in and develop supporting arguments and evidence for. So they then do the experiments or calculations (including consulting other scientists and peer reviewers) to establish a justification and other details. Typically, the evidence they generate or the criticism they receive may not be quite what they expected, so they then revise the claim to be correct, or change it to be a more realistic claim still as yet to be fully substantiated, or they could delete it if it just turns out to be a mess, or they could develop some altogether much better approach to the work as a result of the exploration. 

Typically, many ideas in the paper will be linked to data from conventional experiments. For example a computational paper might calculate that the statistical power of an experiment is too low (there is an unacceptable risk of committing Type II errors), so the authors will decide to improve the experiments. 

If there is code in the paper, every time it is typeset, the code will be run. Therefore the authors of the paper proof-reading the results of the code will have opportunities to try to debug and improve the code. Again, the paper itself is acts like a laboratory, helping the authors refine the science.

Note that systems capable of handling computational papers (including \LaTeX) can create conditional documents: for example, there could be a flag \texttt{publish}. If \texttt{publish} is false, the author can see all their private work and thinking, including all their experimental thinking and workings; but if \texttt{publish} is true, the paper would be typeset as for a submitted paper with the detailed working concealed to ensure a clean and concise presentation. In principle, also submitting to the journal with \texttt{publish} false (hence with data and workings visible) could satisfy journals' code and data requirements. Of course, when there is a large body of data or code, they need not all be an explicit in the computational paper: they would be made available separately in the usual way.

The more we view the paper as a proactive scientific laboratory, the more we gain from the \RAPstar\ perspective, and the more science gains from improved, conceptually broadened, reliable, and reproducible science. The more we will want to engage mature Software Engineering standards too, because the quality and creativity of future science relies on them.

\section{Call to action}\label{summary}
Computer programs are the laboratories of modern scientists, and should be used with a comparable level of care that virologists use in their laboratories --- lab books and all \cite{notebooks} --- and for the same reasons: computer bugs accidentally cultured in one laboratory can infect research and policy worldwide. 

Inadequate code can be extremely problematic. Incorrect results might be used for supporting science, modeling COVID-19, informing public health policy, medical research, or used in other \emph{critical software}. Professional critical software development, as used in critical industries such as aviation and the nuclear power, is (put briefly) based on \emph{correct by construction\/} \cite{cbc}, effectively: design it right first time, supported by numerous rigorous techniques, such as Formal Methods, to manage error. Not coincidentally, such are \emph{exactly\/} the right methods to ensure code is both dependable and scrutable, as required for supporting quality science. Conversely, not following these practices undermines the rigor of science.

An analogous situation arises in ethics. 

Misuse of data, exploiting vulnerable people, and not obtaining informed consent are typical ethical problems. Planned research may be ethically unacceptable, but few people have the objectivity and ethical expertise to make sound ethical judgements, particularly when it comes to assessing their own work. National funders, and others, therefore require Ethics Boards to formally review ethical quality. Medical journals will not publish research that has not undergone appropriate formal ethical review. 

Analogously, and supplementing Ethics Boards, it is proposed that Software Engineering Boards would authorize as well as provide advice to guide the implementation of high-quality Software Engineering to support research and publication processes. Just as journals require conflicts of interest statements, data availability statements, and ethics board clearance, we should move to  all scientific papers and funded research being required to include Software Engineering Board support and clearance statements as appropriate. Note that Software Engineers themselves have a code of ethics that applies to \emph{their\/} work \cite{ethics-code}.

Some journals have policies that code is available (see the \supplement), but they should require that code is not just available in principle but \emph{actually works\/} on the relevant data. Ideally, the authors should test a clean deployed build of their code and save the results. Presumably a paper's authors must have run their code successfully on \emph{some\/} data (if any, but see section \ref{on-code-data-publication} in the \supplement) at least once, so preparing the code and data in a way that is reproducible should be a routine and uncontentious part of the rigorous development of  code underpinning \emph{any\/} scientific claims. This requirement is no more unreasonable than requesting good statistics, as discussed earlier. And the solution is the same: relevant experts --- whether statisticians or Software Engineers --- need to be routinely available and engaged with the science. Software Engineering Boards would be a straight forward way of helping achieve this.

There need to be many Software Engineering Boards (SEBs) to ensure convenient access and oversight, potentially at least one per university. Active, professional Software Engineers should be on these SEBs; this is not a job for people who are not qualified and experienced in the area or who are not actively connected with the true state of the art. There are many high-quality software companies (especially those in safety-critical areas like aviation and nuclear power) who would be willing and competent to help.

Open-source generally improves the quality of software. SEBs will take account of the fact that open-source software enables contributors to be located anywhere, typically without a formal contractual relationship with the leading researchers. Where appropriate, then, SEBs might require \emph{local\/} version control, unit testing, static analysis and other quality control methods for which the lead scientist and Software Engineer remain responsible, and may even need to sign off (and certainly be signed off by the SEB\@). Software Engineering publishers are already developing rigorous badging initiatives to indicate the level of formal review of the quality of software for use in peer-reviewed publications \cite{acm-artifacts}.

A potential argument against SEBs is that they may become onerous, onerous to run, and onerous to comply with their requirements. A more mature view is that SEBs need their processes to be adaptable and proportionate; indeed, few people consider Ethics Boards to be disproportionately onerous. If software being developed is of low risk, then less stringent engineering is required than if the software could cause frequent and critical outcomes, say in their impact on public health policy for a nation. Hence SEBs processes are likely to follow a risk analysis, perhaps starting with a simple checklist. There are standard ways to do this, such as following IEC 61508:2010 \cite{redmill,iec61508} or similar. Following a risk analysis (based on safety assurance cases, controlled documents and so on, if appropriate to the domain), the Board would focus scrutiny where it is beneficial without obstructing routine science.

A professional organization, such as the UK Royal Academy of Engineering ideally working in collaboration with other national international bodies such as IFIP, should be asked to develop and support a framework for SEBs. SEBs could be quickly established to provide direct access to mature Software Engineering expertise for both researchers and for journals seeking competent peer-reviewers. In addition, particularly during a pandemic, SEBs would provide direct access to their expertise for Governments and public policy organizations. Given the urgency, this paper recommends that \emph{ad hoc\/} SEBs should be established for this purpose.

SEBs are a new suggestion, providing a supportive, collaborative process. They meet Tony Hoare's comments about the value of rigorous management of procedures \cite{hoare-unreasonable}, and widen them to non-programmer scientists. Methodological suggestions already made in the literature include open-source and specific Software Engineering methodologies to improve reproducibility \cite{basic-reproducibilty,open-source}. Reference \cite{ABCs-SE} provides an conceptual framework. However, there is scope for further research to provide an evidence base to motivate and assess appropriate interventions (such as those proposed in this paper) to help scientists do more rigorous and effective Software Engineering to support their research and publishing. 

An analogous proposal to SEBs has been made for Methods Review Boards \cite{mrb}, to help scientists ensure the methods they use are appropriate for addressing their research questions. The Methods Review Boards was motivated by an Ethics Board member noticing that often experimental methodologies are inadequate, and will waste time that cannot be corrected until the flaws are spotted too late during peer-review. The paper \cite{mrb} raises many of the same trade-offs that SEBs also face; indeed one would hope that Methods Review Boards would include Software Engineers or have SEBs as sub-boards --- as this paper has argued, Software Engineering is now a key methodology of science. As with SEBs, the goal is not to gatekeep, but to improve.  

\subsection{Action must be interdisciplinary}
\setcounter{footnote}{1}
Code is only part of science, and only one critical factor in the wider reproducibility crisis: SEBs must work with --- and be engaged by --- other reproducibility initiatives, such as TOP and Methods Review Boards.
 
Relying on SEBs \emph{alone\/} would continue one of the current besetting problems about the role of code. The conventional view is that scientists do the hard work compared to the ``easy'' coding work (sections \ref{deceptive-simplicity-of-code} \& \ref{central-role-of-code}) so they just need to tell programmers what to do. This is the view expressed by Landauer in his classic book \emph{The Trouble with Computers\/} \cite{landauer,thimbleby-landauer}, where he argues that the trouble with computers, which he explores at some length, is that we need to spend more effort in working out what computers should do (i.e., do the science better) and then \emph{just\/} tell programmers to do \emph{that}.  

On the contrary, competent Software Engineers have insights into the logic, coherence, complexity, and computability of what they are asked to do, and how that needs refining or optimizing. In other words, the Software Engineers can bring important insights into the science, hence improving or changing the questions and assumptions the science relies on. This insight was widely recognized in the specialist area of numerical computation: ``here is a formula I want you to just code up''~\ldots\ ``but it's ill-conditioned, there is no good answer to that question.'' Ideally, then, it is not  a simple sequential process 

\begin{center}\sf\fbox{science specifies} $\rightarrow$ \fbox{code up} $\rightarrow$ \fbox{get results}, 
\end{center}

\noindent but an iterative cycle of collaboration and growing understanding, informed by Software Engineering best practice (via SEBs), and implemented using papers as laboratories.

In short, the way SEBs work and are used will be crucial to the success of the science they support. Software engineers can help improve the science, so it is not just a matter of asking a SEB whether some coding practices (like documentation) are satisfactory, but whether the SEB has insights into the science itself too. Most effectively, this requires interdisciplinary working practices (science plus Software Engineering) with mutual respect for their contributing expertise.

\subsection{Methodological statements}
Many science journals, conferences, workshops, videos, books, etc, and funders require explicit statements how the authors have conformed to appropriate methodological standards covering issues such as conflicts of interest, ethics, data access, consent, authoring, funders and other acknowledgements of support, and so forth. Conformance to PRISMA, discussed above in section \ref{PRISMA-statement}, is one such methodological statement; like many journals, the high-profile journal \emph{PLOS ONE\/} has a list of similar methodologies it expects authors to comply with as appropriate.

It would be easy for journals and funders to require equivalent types of statements on the quality of code, that is on the quality of its Software Engineering.

Studies of data access statements show that they are unreliable: some authors withdraw papers when journals request access statements \cite{no-raw-data} (indicating that journals that do not make an explicit request are likely publishing papers that will not provide access), and some authors do not respond to access requests \cite{data-access}. How we help scientists who do not want to provide data or code access is one problem, but the serious issue for science is how to ensure any statements made are accurate, and any access provided is actually helpful (e.g., well-defined, versioned, etc) to support useful reproduction. 

Journals and funders should provide support for hosting data and code (and any other relevant data, such as qualitative data, video, etc), and the review process must check that authors actually provide the material as they claim in the methodological and access statements. Conversely, scientists should be able to access funding to ensure data and code access, and, as appropriate, funding for on-going maintenance of the databases, which will typically require funding beyond the end of the normal funding period.

Intellectual Property (IP) is an increasing concern, both for author scientists and their sponsors who want royalties, and for other scientists wishing to freely build on the published science. Particularly concerning access to code, IP is potentially and often is in conflict with scientific openness. Methodological statements should be made concerning any IP associated with code, and to what extent this interferes with open access to code. More routine discussion should include raising known system dependencies, such as operating system, compiler, or special hardware dependencies; it is also appropriate to mention standards conformance, such as to IEEE Floating-Point Arithmetic (IEEE~754).

Journal policies could start to explicitly encourage computationally reproducible science using RAP and \RAPstar\ techniques. That is, the research's methodology itself may be a mixture of data and code. As this paper's \supplement\ shows in section \ref{restrictive-policies}, many journals (e.g., \emph{PLOS ONE\/} and, ironically, \emph{IEEE Transactions on Software Engineering\/}) and repositories have policies that make RAP much harder or just counter-productive at the last step. 

Methodological statements should be required that make clear what access rights are available for RAP or \RAPstar\ material, as it is much more likely to raise IP issues that normal disclosures. In particular, if the authors plan on publishing a series of papers based on the same methodologies, the RAP/\RAPstar\ access might be provided in a later paper or held under escrow by the journal or funding body.

Journals and funders often require data and code access statements, but as this paper has made clear, code is complex and it is rarely easy to understand and scrutinize even with access to substantial documentation (which is unusual). It is therefore recommended that journals and funders require \emph{assurance arguments} \cite{assurance-case}, a familiar technique from the safety assurance domain. Assurance arguments provide a concise, high-level argument that the system does what is claimed. Assurance arguments can be more or less detailed, and more or less formal in their approach; we envisage referees would have views on the level of detail and formality required for any specific contribution.

Finally, as there is no practical distinction between data and code (see \supplement) and methodology (thanks to RAP), and certainly no distinctions that cannot be circumvented, journal and funder policies of code and data access should be reviewed \emph{and unified\/} so that the access statements apply to all information, regardless of any arbitrary classification of it as code or data (or documentation, assurance case, etc).

\subsection{Benefits beyond science}\label{benefits-beyond-science}
Science increasingly recognizes the key supporting roles of code and computation, but many fields do not recognize computation as such, as a skilled discipline, and therefore they are missing out on the leverage that comes with recognizing computation as a first class player in their  activities. 

Healthcare is supported by computers, yet medical research papers remain traditional and do not refer to code. Yet clinical practice relies on computer analysis, so inevitably must use code unrelated to the code developed in research, which is unlikely to be developed to Software Engineering standards. Conversely, the critical issues (including patient safety) that assume code is more reliable than that needed for scientific research do not get evaluated by researchers. The gap is wide. The problems and missed opportunities of under-valued and poorly-managed code are ubiquitous in healthcare \cite{fixit}. Lasting solutions might well be initiated through SEBs or equivalent.

Numerous problems in finance have been facilitated if not precipitated by computational na\" \i vety. JP Morgan Chase (JPM) lost over \$6 billion in a credit derivatives trade \cite{jpm}, in a costly parody of bad science --- compare the discussion here with figure \ref{excelFigure}, in a series of figures in the \supplement, which illustrates the same problem as it presents in many scientific papers. 

As reported in \cite{jpm}, the traders did not understand the trades, did not monitor them, doubled down when results were poor, and did not communicate the extent of their losses. They were using manual coding methods; as the report says: 

\begin{quote}\raggedright
{\sf\setbox0=\hbox{``}
\hskip -\wd0\box0
[\ldots] the model operated through a series of Excel spreadsheets, which had to be completed manually, \textbf{by a process of copying and pasting data from one spreadsheet to another}. [Our emphasis.]

\hskip -\wd0\box0
[\ldots] this individual immediately made certain adjustments to formulas in the spreadsheets he used. These changes, which were not subject to an appropriate vetting process, inadvertently introduced two calculation errors 

\hskip -\wd0\box0
[\ldots] after subtracting the old rate from the new rate, the spreadsheet divided by their sum instead of their average, as the modeler had intended.'' 

etc}
\cite{jpm-report}

\end{quote}

The report does not detail how the Excel spreadsheets were specified or coded, seemingly as unawqre of Software Engineering as the traders. However, one infers from the brief discussion of data handling (which is easier to automate) that it was all an unconsciously incompetent process.

Reviewers in JPM failed to scrutinize not just the coding, but the trades informed by the code. They passed on optimistic reports. Then there was a merry-go-round of blame: ``the information communicated to the Risk Policy Committee \ldots\ did not suggest any significant problems \ldots\  there was no robust debate with the right facts at the right level about the portfolio risk.'' UK and US governments are now investigating fraud. Again, lasting solutions might well be initiated through SEBs or equivalent.

Without taking the lessons of improving Software Engineering to other fields, including improving and broadening the recognition and career paths for developers, there will continue to be an unfortunate and unnecessary disconnect between competent coding and actual practice. Everything, from healthcare to finance --- not just science --- will suffer because the critical contributions of dependable code, quality Software Engineering, and competent Computational Thinking are not recognized, understood, valued, or required. 

\ignore{\subsection{Suggestions for further work}
Although this study of Software Engineering in scientific research (specifically, in peer-reviewed publications) was necessarily interpretive in nature, the corpus of data (namely, the selected \the\dataN\ papers) was rigorously gathered so that it could be later updated or extended in size or scope. However, the insights appear to be general.

Further work to extend the scope of the survey beyond the basic requirements of the present paper is of course worthwhile, but the following cases (listed in the next few paragraphs) suggest that the problem is widespread. We argue, then, that we should be focusing effort on avoiding problems and considering proposed solutions (see section \ref{summary}), not just assessing the problems highlighted in this paper with increasing scale or rigor. }

\section{Conclusions}
In a pandemic, epidemiological modeling (discussed in this paper), track and trace \cite{excel-fiasco}, modeling mutation pressures against vaccine shortages and delays between vaccinations \cite{science-delays}, and so on, drive public policy and have a direct impact, whether positively or negatively, on the quality of life. 

While this paper was originally motivated by Ferguson's public statements \citeeg{tweet,ferguson-interview} about his high-profile COVID modeling, the wider evidence reviewed here suggests that scientific coding practice is inadequate in every field. As science becomes more and more reliant on computers, we need to correspondingly improve the quality of code, the quality of code policies, the quality of Software Engineering, and the quality of all scientists' understanding of computation and how to manage its unlimited complexity. 

The main challenges to mature computationally-realistic science are:

\enum=1
\begin{enumerate}\raggedright
\item 
To manage software development to reduce the unnoticed and unknown impacts of bugs and poor programming practices that research and publications rely on. Computer code should be explicit, accessible (well-structured, etc), and adequately documented. Papers should be explicit on their software methodologies, limitations and weaknesses, just as Whitty expressed more generally about the standards of science \cite{whitty}. Professional software methodologies should not be ignored.

\item 
To use computation to help make scientific processes explicit, so that they can be reproduced, scrutinized and improved. RAP is an increasingly popular way to help do some of this, but as this paper points out, RAP should be generalized to \RAPstar\ to help the computational parts of science as well, leading to a virtuous circle.

\item
To support and develop the scientific community in the professional use of computation.

\item
To find effective ways to promote professional software engineers being recognized and participating  fully in scientific research, like professional statisticians routinely support quality research.
\end{enumerate}

While programming seems easy, and is often taken for granted and done casually, programming \emph{well\/} is very difficult \cite{fixit}. We know from software research that ordinary programming is very buggy and unreliable. Without adequately specified and documented code and data, research is not open to scrutiny, let alone proper review, and its quality is suspect. Some have argued that availability of code and data ensure research is reproducible, but that is na\"\i ve criterion: computer programs are easy to run and reproduce results, but being able to reproduce something of low quality does not magically make it more reliable \cite{reproducibility,relit,popper-conjectures-refutations}. 

Software Engineering Boards, as proposed in this paper, are an initial, straightforward, constructive, and practical way to support and improve computer-based science. 

This paper's \supplement\ summarizes relevant Software Engineering good practice that Software Engineering Boards would draw on, including discussing how and why Software Engineering helps improve code reliability, dependability, and quality.

\makeatletter
\immediate\write\@auxout{\string\def\string\continueSectionNumbers{\arabic{section}}}
\makeatother

\section*{Supporting information}
\newcount\csrefcount \csrefcount=0
\def\csref{\global\advance\csrefcount by 1}
\long\def\ethics#1{\paragraph*{Ethics}#1}
\long\def\ack#1{\paragraph*{Acknowledgments}#1}
\long\def\dataaccess#1{\paragraph*{Data and code access}#1}
\long\def\aucontribute#1{\paragraph*{Author contribution}#1}
\long\def\competing#1{\paragraph*{Competing interests}#1}
\long\def\funding#1{\paragraph*{Funding}#1}

\ack{The author is very grateful for comments from: \csref Ross Anderson, \csref Nicholas Beale, \csref Ann Blandford, \csref Paul Cairns, \csref Rod Chapman, \csref Jos\'e Corr\'ea de~S\`a, \csref Paul Curzon, \csref Jeremy Gibbons, \csref Richard Harvey, \csref Will Hawkins, \csref Konrad Hinsen, \csref Ben Hocking, \csref Daniel Jackson, \csref Peter Ladkin, \csref Bev Littlewood, \csref Paolo Masci, Stephen Mason, \csref Robert Nachbar, \csref Martin Newby, \csref Patrick Oladimeji, \csref Claudia Pagliari, \csref Simon Robinson, \csref Jonathan Rowanhill, \csref John Rushby, \csref Susan Stepney, \csref Isaac Thimbleby, \csref Prue Thimbleby, \csref Will Thimbleby, \csref Martyn Thomas, and \csref Ben Wilson. The author also thanks the anonymous \emph{Computer Journal\/} referees who also contributed to the quality of this paper.}
%\the\csrefcount\ computer science reviewers.

%\ethics{This article presents research with ethical considerations but does not fall within the usual scope of ethics policies.}

\dataaccess{There is an extended discussion of the methodology of this paper in the \supplement, section \ref{on-code-data-publication}. The \supplement\ also presents all raw data in tabular form. All material is also available for download at \url{github.com/haroldthimbleby/Software-Enginering-Boards},\footnote{\color{red}This is a temporary URL before meeting publication repository requirements for accepted papers.\color{black}} which has been tested in a clean build, etc. 

The data is encoded in JSON\@. JavaScript code, conveniently in the same file as the JSON data, checks (with \the\JSONerrorCount\ possible classes of error report) and converts the JSON data into \LaTeX\ number registers and summary tables, etc, thus making it trivial to typeset all results reliably in this paper and in its \supplement\ \emph{directly\/} from the automatic data analysis.

In addition, a standard CSV file is generated from the JSON in case this is more convenient, for instance to browse directly as a spreadsheet or to import easily into other programs.}

\aucontribute{Harold Thimbleby is the sole author. An preliminary outline of this paper, containing no supplementary material or data, was submitted to the UK Parliamentary Science and Technology Select Committee's inquiry into UK Science, Research and Technology Capability and Influence in Global Disease Outbreaks, under reference LAS905222, 7 April, 2020. The evidence, which was not peer-reviewed and is only available after an explicit search, briefly summarizes the case for Software Engineering Boards, but without the detailed analysis and case studies of the literature, etc, that are in the present paper. It is available to the public \cite{parliamentary-evidence,my-parliamentary-evidence}.}

\competing{The author declares no competing interests.}

\funding{This work was jointly supported by See Change (M\&RA-P), Scotland (an anonymous funder), by the Engineering and Physical Sciences Research Council [grant EP/M022722/1], by the Royal Academy of Engineering through the Engineering~X Pandemic Preparedness Programme [grant EXPP2021\textbackslash 1\textbackslash 186], and by Assuring Autonomy International Programme, Assuring Safe AI in Ambulance Service Triage. The funders had no involvement in the research or in this paper.}

%\nolinenumbers

{\raggedright

\bibliographystyle{revisedCompJ}
\def\urlprefix{}
\initialiseBibliography{References}{1}{}

\newcount \bibitemn \bibitemn = 0
\newcount \bibnewcolumn 

\bibnewcolumn = 112 % bibentry number of last entry on (final) left hand column
% Hint: set it to a very large number, so the final column is as unbalanced as it gets
% then, if it's worth fixing the final page, lower the number to some citation on the 
% lower part of the left hand column; this citation will then become the last in that
% column. Hopefully the righthand column will be of equal length!
% PS Don't forget to check there is nothing on the following page!

% if you can't get a decent column balance, tinker with the inter-citation spacing, below
% it currently looks like "\vskip .2ex plus 1ex minus 1ex"
% you'll be surprised what adjusting it by only 0.1ex or less achieves after 100 citations!

\let\oldbibitem=\bibitem

% calculate bib entry width
\newdimen \bibEntryWidth
\bibEntryWidth = \columnwidth
\advance \bibEntryWidth by -\columnsep

% I haven't figured out how to make the \labelwidth etc variable - it'd be nice
% to have it suitable for numbers up to 9, then for numbers up to 99, then up to 999
% depending on the largest bibitem number width in the column. As it is, it's set
% to the width of 999 for all bibitems even though it takes a while for bibitems to get 
% up to 3 figures. Sigh

\def\bibitem#1{%
\ifnum \bibitemn = \bibnewcolumn \egroup\newpage\bgroup \fi
\ifnum \bibitemn = 0 \else \egroup \fi
\advance \bibitemn by 1
\vskip .3ex plus 1ex minus 1ex 
\leavevmode \vbox\bgroup \parshape 1 0in \bibEntryWidth 
\oldbibitem{#1}}

\def\endofbibliography{ \egroup \vfill}

% the class comjnl has a minor formatting bug in the setting of labelwidth; fixed below:
\makeatletter
\def\thebibliography#1{\section*{References}\small\list
 {[\arabic{enumi}]}{\labelsep 0.5em%
\hyphenpenalty=50\itemsep=-1pt
 \settowidth\labelwidth{[0#1]}% previously too small, so added an extra digit (0)
 \leftmargin\labelwidth
\advance\leftmargin\labelsep
 \usecounter{enumi}}
 \def\newblock{\hskip .14em plus .38em minus -.1em}
 \sloppy
 \sfcode`\.=1000\relax}
\let\endthebibliography=\endlist

\def\@cite#1#2{[{#1\if@tempswa , #2\fi}]}
\def\@biblabel#1{[#1]}
\makeatother

\bibliography{paper-seb-main.bib}

% Note that the supplementary material calculates how many references there are,
% because it reads in this .tex file's .aux file so that it can cite the references 
% using the same number sequence. 
%
% It simply redefines \bibitem to work it out while it
% defines all the citations from the bibliography above.
%
% The same technique allows the supplementary material to have 2 more bibliographies,
% with consecutive numbering.
}

%\label{LastPage}
%save Latex register    {linenumber}
%\makeatletter
%\immediate\write\@auxout{\string\def\string\MainPaperLastLine{\arabic{linenumber}}}
%\makeatother
\end{document}

\def \allInOne {define this so we don't include header in supplemental material}
\input paper-seb-supplementary-material.tex
\end{document}

%\immediate\write\@auxout{\string \newcount \string \savedlastReferenceNumber
%\string \savedlastReferenceNumber=\the\savedlastReferenceNumber}

\end{document} 
